vm/systemSmartContracts/esdt.go

```diff
@@ -795,6 +805,206 @@ func (e *esdt) controlChanges(args *vmcommon.ContractCallInput) vmcommon.ReturnC
+
+	address := args.Arguments[1]
+	esdtRole, isNew := getRolesForAddress(token, address)
+	if isNew {
+		e.eei.AddReturnMessage("address does not have special role")
+		return vmcommon.UserError
+	}
+
+	for _, arg := range args.Arguments[2:] {
+		index, exist := doesRoleExist(esdtRole, arg)
+		if !exist {
+			e.eei.AddReturnMessage("role does not exist for address")
+			return vmcommon.UserError
+		}
+
+		copy(esdtRole.Roles[index:], esdtRole.Roles[index+1:])
+		esdtRole.Roles[len(esdtRole.Roles)-1] = nil
+		esdtRole.Roles = esdtRole.Roles[:len(esdtRole.Roles)-1]
+	}
+
+	err := e.sendRoleChangeData(args, core.BuiltInFunctionUnSetESDTRole)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = e.saveToken(args.Arguments[0], token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) sendRoleChangeData(args *vmcommon.ContractCallInput, builtInFunc string) error {
+	esdtSetRoleData := builtInFunc + "@" + hex.EncodeToString(args.Arguments[0])
+	for _, arg := range args.Arguments[2:] {
+		esdtSetRoleData += "@" + hex.EncodeToString(arg)
+	}
+
+	err := e.eei.Transfer(args.Arguments[1], e.eSDTSCAddress, big.NewInt(0), []byte(esdtSetRoleData), 0)
+	return err
+}
+
+func (e *esdt) getAllAddressesAndRoles(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) != 1 {
+		e.eei.AddReturnMessage("needs 1 argument")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		e.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.UserError
+	}
+
+	token, err := e.getExistingToken(args.Arguments[0])
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	for _, esdtRole := range token.SpecialRoles {
+		e.eei.Finish(esdtRole.Address)
+		for _, role := range esdtRole.Roles {
+			e.eei.Finish(role)
+		}
+	}
+
+	return vmcommon.Ok
+}
+
+func checkDuplicates(arguments [][]byte) error {
+	mapArgs := make(map[string]struct{})
+	for _, arg := range arguments {
+		_, found := mapArgs[string(arg)]
+		if !found {
+			mapArgs[string(arg)] = struct{}{}
+		}
+		return vm.ErrDuplicatesFoundInArguments
+	}
+
+	return nil
+}
+
+func validateRoles(arguments [][]byte) error {
+	err := checkDuplicates(arguments)
+	if err != nil {
+		return err
+	}
+
+	for _, arg := range arguments {
+		err = isSpecialRoleValid(string(arg))
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isSpecialRoleValid(argument string) error {
+	switch argument {
+	case core.ESDTRoleLocalMint:
+		return nil
+	case core.ESDTRoleLocalBurn:
+		return nil
+	default:
+		return vm.ErrInvalidArgument
+	}
+}
+
+func doesRoleExist(esdtRoles *ESDTRoles, role []byte) (int, bool) {
+	for i, currentRole := range esdtRoles.Roles {
+		if bytes.Equal(currentRole, role) {
+			return i, true
+		}
+	}
+	return -1, false
+}
+
+func getRolesForAddress(token *ESDTData, address []byte) (*ESDTRoles, bool) {
```

Comment:
Would have preferred a single output argument as that argument would have provided info that the bool variable is actually "is new" flag. Otherwise we have to continuously come back to this function to check what does the bool flag means.

Reply:
this is used in multiple places like this in staking / validator.go